README
= Ruby Graph Library (RGL)

RGL is a framework for graph data structures and algorithms.

The design of the library is much influenced by the Boost Graph Library (BGL)
which is written in C++ heavily using its template mechanism. Refer to
http://www.boost.org/libs/graph/doc for further links and documentation on graph
data structures and algorithms and the design rationales of BGL.

A comprehensive summary of graph terminology can be found in the the graph
section of the <em>Dictionary of Algorithms and Data Structures</em> at
http://www.nist.gov/dads/HTML/graph.html.

== Design principles

This document concentrates on the special issues of the implementation in
Ruby. The main design goals directly taken from the BGL design are:

* An interface for how the structure of a graph can be accessed using a generic
  interface that hides the details of the graph data structure
  implementation. This interface is defined by the module Graph, which should be
  included in concrete classes.

* A standardized generic interface for traversing graphs (RGL::GraphIterator)

RGL provides some general purpose graph classes that conform to this interface,
but they are not meant to be the *only* graph classes. As in BGL I believe that
the main contribution of the RGL is the formulation of this interface.

The BGL graph interface and graph components are generic in the sense of the C++
Standard Template Library (STL). In Ruby other techniques are available to
express the generic character of the algorithms and data structures mainly using
mixins and iterators. The BGL documentation mentions three means to achieve
genericity:

* Algorithm/Data-Structure Interoperability
* Extension through Function Objects and Visitors
* Element Type Parameterization
* Vertex and Edge Property Multi-Parameterization

The first is easily achieved in RGL using mixins, which of course is not as
efficient than C++ templates (but much more readable :-). The second one is
even more easily implemented using standard iterators with blocks or using the
Stream[http://rgl.rubyforge.org/stream/files/README.html] module. The third one
is no issue since Ruby is dynamically typed: Each object can be a graph
vertex. There is no need for a vertex (or even edge type). In the current
version of RGL properties of vertices are simply attached using hashes. At
first there seems to be not much need for the graph property machinery.

=== Algorithms

The first version of RGL only contains a core set of algorithm patterns:

* Breadth First Search (RGL::BFSIterator)
* Depth First Search (RGL::DFSIterator)

The algorithm patterns by themselves do not compute any meaningful quantities
over graphs, they are merely building blocks for constructing graph
algorithms. The graph algorithms in RGL currently include:

* Topological Sort (RGL::TopsortIterator)
* Connected Components (RGL::Graph#each_connected_component)
* Strongly Connected Components (RGL::Graph#strongly_connected_components)
* Transitive Closure (RGL::Graph#transitive_closure)

=== Data Structures

RGL currently provides two graph classes that implement a generalized adjacency
list and an edge list adaptor.

* RGL::AdjacencyGraph
* RGL::ImplicitGraph

The AdjacencyGraph class is the general purpose *swiss army knife* of graph
classes. It is highly parameterized so that it can be optimized for different
situations: the graph is directed or undirected, allow or disallow parallel
edges, efficient access to just the out-edges, fast vertex insertion and removal
at the cost of extra space overhead, etc.

=== Differences to BGL

The concepts of IncidenceGraph, AdjacencyGraph and VertexListGraph (see
http://www.boost.org/libs/graph/doc/IncidenceGraph.html) are here bundled in the
base graph module. Most methods of IncidenceGraph should be standard in the base
module Graph. The complexity guarantees can not necessarily provided.  See
http://www.boost.org/libs/graph/doc/graph_concepts.html.

== Installation

RGL is depended on the
stream[http://rgl.rubyforge.org/stream/files/README.html] library which can
also be downloaded from http://rubyforge.org/frs/?group_id=110. If you use gem
to install RGL the stream library will be installed as a prerequisite.

=== GEM Installation

Download the GEM file and install it with ..

  % gem install rgl-VERSION.gem

or directly with 

  % gem install rgl

Use the correct version number for VERSION (e.g. 0.2.x).  You may need root
privileges to install.

=== Running tests

RGL comes with a Rakefile which automatically runs the tests. Goto the
installation directory and start rake:

 % gem env
 Rubygems Environment:
  - VERSION: 0.9.0 (0.9.0)
  - INSTALLATION DIRECTORY: /usr/lib/ruby/gems/1.8
  - GEM PATH:
     - /usr/lib/ruby/gems/1.8
  - REMOTE SOURCES:
     - http://gems.rubyforge.org

 % cd /usr/lib/ruby/gems/1.8/gems/rgl-0.3.0/
 % rake
 (in /usr/lib/ruby/gems/1.8/gems/rgl-0.3.0)
 /usr/bin/ruby1.8 -Ilib:tests "/usr/lib/ruby/gems/1.8/gems/rake-0.7.3/lib/rake/rake_test_loader.rb" "tests/TestTransitiveClosure.rb" "tests/TestComponents.rb" "tests/TestCycles.rb" "tests/TestDirectedGraph.rb" "tests/TestEdge.rb" "tests/TestGraph.rb" "tests/TestGraphXML.rb" "tests/TestImplicit.rb" "tests/TestUnDirectedGraph.rb" "tests/TestTraversal.rb" "tests/TestDot.rb" "tests/TestRdot.rb" 
 Loaded suite /usr/lib/ruby/gems/1.8/gems/rake-0.7.3/lib/rake/rake_test_loader
 Started
 ......................................................................................................................................................
 Finished in 0.750958 seconds.

 86 tests, 625 assertions, 0 failures, 0 errors
 
=== Code coverage

Running rcov[http://eigenclass.org/hiki.rb?rcov] on the test suite generates this[link:coverage/index.html] result.

=== Normal Installation

You have to install stream library before. You can than install RGL with the
following command:

  % ruby install.rb

from its distribution directory. To uninstall it use

  % ruby install.rb -u

== Example irb session with RGL

  irb> require 'rgl/adjacency'
  irb> dg=RGL::DirectedAdjacencyGraph[1,2 ,2,3 ,2,4, 4,5, 6,4, 1,6]
  # Use DOT to visualize this graph:
  irb> require 'rgl/dot'
  irb> dg.write_to_graphic_file('jpg')
  "graph.jpg"
  
The result: link:../examples/example.jpg  
  
  irb> dg.directed?
  true
  irb> dg.vertices
  [5, 6, 1, 2, 3, 4]
  irb> dg.has_vertex? 4
  true

Every object could be a vertex (there is no class Vertex), even the class
object _Object_:

  irb> dg.has_vertex? Object
  false
  irb> dg.edges.sort.to_s
  "(1-2)(1-6)(2-3)(2-4)(4-5)(6-4)"
  irb> dg.to_undirected.edges.sort.to_s
  "(1=2)(1=6)(2=3)(2=4)(5=4)(6=4)"

Add inverse edge (4-2) to directed graph:

  irb> dg.add_edge 4,2

(4-2) == (2-4) in the undirected graph:

  irb> dg.to_undirected.edges.sort.to_s
  "(1=2)(1=6)(2=3)(2=4)(5=4)(6=4)"

(4-2) != (2-4) in directed graphs:

  irb> dg.edges.sort.to_s
  "(1-2)(1-6)(2-3)(2-4)(4-2)(4-5)(6-4)"
  irb> dg.remove_edge 4,2
  true

<em>Topological sort</em> is realized with as iterator:

  require 'rgl/topsort'
  irb> dg.topsort_iterator.to_a
  [1, 2, 3, 6, 4, 5]

A more elaborated example showing <em>implicit graphs</em>:

  def module_graph
	RGL::ImplicitGraph.new { |g|
	  g.vertex_iterator { |b|
		ObjectSpace.each_object(Module, &b) 
	  }
	  g.adjacent_iterator { |x, b|
		x.ancestors.each { |y|
		  b.call(y) unless x == y || y == Kernel || y == Object
		}
	  }
	  g.directed = true
	}
  end

This function creates a directed graph, with vertices being all loaded modules:

  g = module_graph

We only want to see the ancestors of RGL::AdjacencyGraph:

  tree = bfs_search_tree_from(g,RGL::AdjacencyGraph)

Now we want to visualize this component of g with DOT.  We therefore create a
subgraph of the original graph, using a filtered graph:

  g = g.vertices_filtered_by {|v| tree.has_vertex? v}

Create the graphics with DOT:

  g.write_to_graphic_file

produces module_graph.jpg: link:../examples/module_graph.jpg

Look for more in the examples directory (i.e.
examples.rb[link:files/examples/examples_rb.html]).

== My del.icio.us links concerning RGL

I collect some links to stuff around RGL at http://del.icio.us/monora/rgl. I
registered RGL at SWiK[http://swik.net/rgl].

== Credits

Many thanks to Robert Feldt which also worked on a graph library
(http://rockit.sf.net/subprojects/graphr) who pointed me to BGL and many other
graph resources.

Robert kindly allowed to integrate his work on graphr, which I did not yet
succeed. Especially his work to output graphs for
GraphViz[http://www.research.att.com/sw/tools/graphviz/download.html] is much
more elaborated than the minimal support in dot.rb.

Jeremy Siek one of the authors of the nice book "The Boost Graph Library (BGL)"
(http://www.boost.org/libs/graph/doc) kindly allowed to use the
BGL documentation as a _cheap_ reference for RGL. He and Robert also gave
feedback and many ideas for RGL.

Dave Thomas for RDoc[http://rdoc.sourceforge.net] which generated what you read
and matz for Ruby. Dave included in the latest version of RDoc (alpha9) the
module dot/dot.rb which I use instead of Roberts module to visualize graphs
(see rgl/dot.rb).

Jeremy Bopp, John Carter, Sascha Doerdelmann, Shawn Garbett and Andreas Schörk
for contributing additions, test cases and bugfixes.

Kirill Lashuk who started to take over further development in November 2012.

== Copying

RGL is Copyright (c) 2002,2004,2005,2008 by Horst Duchene.  It is free software, and may be
redistributed under the terms specified in the README file of the Ruby distribution.

== Support

Please contact me at mailto:monora@gmail.com with bug reports
suggestions, and other comments. If you send patches, it would help if
they were in-line (not attachments) and generated using "diff -u".




ChangeLog
2022-08 Release 0.5.9

Dan Čermák
 * Drop lazy priority queue (#64) (3b1db1)

2022-06 Release 0.5.8

Horst Duchene
 * switch to github actions (56030d)

2020-12 Release 0.5.7

Horst Duchene
 * Fully automate dev setup with Gitpod (41dd00)
 * Add Dockerfile to install graphviz (2bd738)
 * Examples do not call dotty (6bba96)
 * Add ruby license file (a21aa5)
ujihisa <ujihisa@users.noreply.github.com>
 * Test against Ruby 2.6 and 2.7 as well (50ac7c)
 * Fix dead links (9184f3)
Harry Lascelles <hlascelles@users.noreply.github.com>
 * Update .travis.yml (45b9a2)
 * Make the links more explicit (95dc3b)
Harry Lascelles
 * Add explicit license to gemspec (de3647)

2019-01 Release 0.5.6

Artemy Kirienko
 * PR #42 Add method Graph#path?(u, v) to check if a path exists between two vertices
Horst Duchene
 * Fix #47 set_to_begin for graph iterator (881aa8)

2019-01 Release 0.5.4

Lia Skalkos
 * Enable options to be passed to #write_to_graphic_file (4ca972). For details see PR #41
Horst Duchene
 * Fix travis-ci errors
 * Add new ruby versions
 * Fix gemspec errors
 * Fix lint warnings
 * Use version stream 0.5.2

2017-04 Release 0.5.3

Horst Duchene
 * Issue #38: Add error handling or dot functions. (719e38, 5a3423)
Thomas Orozco
 * Remove Enumerable Extension (fde8a5)
 * Update to codeclimate-test-reporter 1.x (25fdb5)
Mario Daskalov
 * Clarify that you need graphviz in the README (35a4b4)

2016-05 Release 0.5.2

Horst Duchene
 * Issue #21: Use new method vertex_id instead of object_id to identify vertices in dot export. (fa7592)
 * Integrate Code Climate's test coverage reporting (0ab722)
 * Clarify traversal order of DFS search (see #20). (afa788)
Chase Gilliam
 * drop 1.9.3 add newer jruby and rubinius (fad333)
Matías Battocchia
 * Switched to a different heap implementation. (bd7c13)
gorn
 * Adding failing test for issue #24 (1f6204)

2015-12 Release 0.5.1

Horst Duchene
 * Changed edge sequence to match example picture (daa88e)

Chase
 * updated algorithms to 6.1 and added test unit to support newer rubies (fbd874)

Louis Rose
 * Fix #15. Use object IDs rather than labels to identify vertexs in DOT graph
   to ensure that distinct nodes that share a label are shown. (33206f, 4fc455)

2014-12 Release 0.5.0

Horst Duchene
  * Changed edge sequence to match example picture (daa88e)
  * Fixed comment (6a6c93)
  * Fixed spelling (7ca281)
Horst Duchêne
Chase
  * updated algorithms to 6.1 and added test unit to support newer rubies (fbd874)
Louis Rose
  * Fix #15. Use object IDs rather than labels to identify vertexs in DOT graph to ensure that distinct nodes that share a label are shown. (33206f)
  * Issue #15. Fix tests. (4fc455)

2014-12 Release 0.5.0
Horst Duchene
  * Changed edge sequence to match example picture (daa88e)
  * Fixed comment (6a6c93)
  * Fixed spelling (7ca281)
Chase
  * updated algorithms to 6.1 and added test unit to support newer rubies (fbd874)
Louis Rose
  * Fix #15. Use object IDs rather than labels to identify vertexs
	in DOT graph to ensure that distinct nodes that share a label are
	shown. (33206f)
  * Issue #15. Fix tests. (4fc455)

2014-12 Release 0.5.0

This release mainly contains the contributions of Kirill, who added many algorithms to the library. Thank you Kirill!

 * @matiaskorhonen: Fixes the image paths in the README (#14)
 * @monora: Implicit graph example fails (#13)
 * @KL-7: Implement Graph#bipartite_sets. (#12)
 * @monora:  syntax error in dot file for undirected graph (#11)
 * @KL-7: Edmonds-Karp algorithm for maximum flow (#10)
 * @KL-7: Prim's algorithm for minimum spanning tree (#9)
 * @carlosantoniodasilva: Run tests on Ruby 2.0 and remove deprecation warning (#8)
 * @KL-7: Bellman-Ford shortest paths algorithm (#7)
 * @KL-7: Minor improvements (asserts) for Dijkstra algorithm (#6)
 * @KL-7: Add Dijkstra shortest path algorithm. (#5)
 * @KL-7: Indentation and whitespaces clean up of examples (#4)
 * @KL-7: Travis configuration and README updates (#3)
 * @KL-7: Code clean up and configuration updates (#2)
 * @aschoerk: Renamed test-directory, (includes Rakefile), fixed TestComponents (#1)

2008-08-27 23:30  javanthropus

	* lib/rgl/base.rb: Preparing for 0.4.0 release

2008-08-26 20:07  javanthropus

	* lib/rgl/dot.rb, lib/rgl/rdot.rb, tests/TestRdot.rb: Move the DOT
	  module into the RGL module

	  * This eliminates a class conflict with the DOT module from rdoc
	  when building RGL's documentation * Also remove the superfluous
	  DOT prefixes from class names in the DOT module

2008-08-24 06:16  javanthropus

	* Rakefile: Remove some comments I accidentally left in while
	  testing rdoc functionality

2008-08-24 06:03  javanthropus

	* Rakefile, lib/rgl/transitiv_closure.rb, lib/rgl/transitivity.rb,
	  tests/TestTransitiveClosure.rb, tests/TestTransitivity.rb:
	  Feature 21641: Added transitive reduction functionality

	  * Updated the gem description to announce this functionality *
	  Moved the transitive closure functionality into the
	  transitivity.rb file along with the transitive reduction
	  funtionality * Modifed the transitiv_closure.rb file to simply
	  load the transitivity.rb file for backward compatibility * Moved
	  all transitivity tests into TestTransitivity.rb

2008-08-23 15:45  javanthropus

	* lib/rgl/condensation.rb, lib/rgl/transitiv_closure.rb,
	  tests/TestTransitiveClosure.rb: Defect 21630: Fixed transitive
	  closure

	  * The fix is based on the algorithm described in the
	  documentation for the implementation of transitive closure in
	  Boost * Along with the fix, performance is improved to O(|V||E|)
	  * This implementation needs graph condensation, so that function

lib/rgl/rdot.rb
module RGL

  # This is a modified version of +dot.rb+ from {https://ruby.github.io/rdoc Dave
  # Thomas's rdoc project}. I renamed it to +rdot.rb+ to avoid collision with an
  # installed rdoc/dot.
  #
  # It also supports undirected edges.
  module DOT

    # options for node declaration

    NODE_OPTS = [
        # attributes due to
        # https://www.graphviz.org/Documentation/dotguide.pdf
        # February 23, 2008
        'color', # default: black; node shape color
        'comment', # any string (format-dependent)
        'distortion', # default: 0.0; node distortion for shape=polygon
        'fillcolor', # default: lightgrey/black; node fill color
        'fixedsize', # default: false; label text has no affect on node size
        'fontcolor', # default: black; type face color
        'fontname', # default: Times-Roman; font family
        'fontsize', #default: 14; point size of label
        'group', # name of node's group
        'height', # default: .5; height in inches
        'label', # default: node name; any string
        'layer', # default: overlay range; all, id or id:id
        'orientation', # dafault: 0.0; node rotation angle
        'peripheries', # shape-dependent number of node boundaries
        'regular', # default: false; force polygon to be regular
        'shape', # default: ellipse; node shape; see Section 2.1 and Appendix E
        'shapefile', # external EPSF or SVG custom shape file
        'sides', # default: 4; number of sides for shape=polygon
        'skew' , # default: 0.0; skewing of node for shape=polygon
        'style', # graphics options, e.g. bold, dotted, filled; cf. Section 2.3
        'URL', # URL associated with node (format-dependent)
        'width', # default: .75; width in inches
        'z', #default: 0.0; z coordinate for VRML output

        # maintained for backward compatibility or rdot internal
        'bottomlabel', # auxiliary label for nodes of shape M*
        'bgcolor',
        'rank',
        'toplabel' # auxiliary label for nodes of shape M*
    ]

    # options for edge declaration

    EDGE_OPTS = [
        'arrowhead', # default: normal; style of arrowhead at head end
        'arrowsize', # default: 1.0; scaling factor for arrowheads
        'arrowtail', # default: normal; style of arrowhead at tail end
        'color', # default: black; edge stroke color
        'comment', # any string (format-dependent)
        'constraint', # default: true use edge to affect node ranking
        'decorate', # if set, draws a line connecting labels with their edges
        'dir', # default: forward; forward, back, both, or none
        'fontcolor', # default: black type face color
        'fontname', # default: Times-Roman; font family
        'fontsize', # default: 14; point size of label
        'headlabel', # label placed near head of edge
        'headport', # n,ne,e,se,s,sw,w,nw
        'headURL', # URL attached to head label if output format is ismap
        'label', # edge label
        'labelangle', # default: -25.0; angle in degrees which head or tail label is rotated off edge
        'labeldistance', # default: 1.0; scaling factor for distance of head or tail label from node
        'labelfloat', # default: false; lessen constraints on edge label placement
        'labelfontcolor', # default: black; type face color for head and tail labels
        'labelfontname', # default: Times-Roman; font family for head and tail labels
        'labelfontsize', # default: 14 point size for head and tail labels
        'layer', # default: overlay range; all, id or id:id
        'lhead', # name of cluster to use as head of edge
        'ltail', # name of cluster to use as tail of edge
        'minlen', # default: 1 minimum rank distance between head and tail
        'samehead', # tag for head node; edge heads with the same tag are merged onto the same port
        'sametail', # tag for tail node; edge tails with the same tag are merged onto the same port
        'style', # graphics options, e.g. bold, dotted, filled; cf. Section 2.3
        'taillabel', # label placed near tail of edge
        'tailport', # n,ne,e,se,s,sw,w,nw
        'tailURL', # URL attached to tail label if output format is ismap
        'weight', # default: 1; integer cost of stretching an edge

        # maintained for backward compatibility or rdot internal
        'id'
    ]

    # options for graph declaration

    GRAPH_OPTS = [
        'bgcolor', # background color for drawing, plus initial fill color
        'center', # default: false; center draing on page
        'clusterrank', # default: local; may be "global" or "none"
        'color', # default: black; for clusters, outline color, and fill color if
                 # fillcolor not defined
        'comment', # any string (format-dependent)
        'compound', # default: false; allow edges between clusters
        'concentrate', # default: false; enables edge concentrators
        'fillcolor', # default: black; cluster fill color
        'fontcolor', # default: black; type face color
        'fontname', # default: Times-Roman; font family
        'fontpath', # list of directories to search for fonts
        'fontsize', # default: 14; point size of label
        'label', # any string
        'labeljust', # default: centered; "l" and "r" for left- and right-justified
                     # cluster labels, respectively
        'labelloc', # default: top; "t" and "b" for top- and bottom-justified
                    # cluster labels, respectively
        'layers', # id:id:id...
        'margin', # default: .5; margin included in page, inches
        'mclimit', # default: 1.0; scale factor for mincross iterations
        'nodesep', # default: .25; separation between nodes, in inches.
        'nslimit', # if set to "f", bounds network simplex iterations by
                   # (f)(number of nodes) when setting x-coordinates
        'nslimit1', # if set to "f", bounds network simplex iterations by
                    # (f)(number of nodes) when ranking nodes
        'ordering', # if "out" out edge order is preserved
        'orientation', # default: portrait; if "rotate" is not used and the value is
                       # "landscape", use landscape orientation
        'page', # unit of pagination, e.g. "8.5,11"
        'rank', # "same", "min", "max", "source", or "sink"
        'rankdir', # default: TB; "LR" (left to right) or "TB" (top to bottom)
        'ranksep', # default: .75; separation between ranks, in inches.
        'ratio', # approximate aspect ratio desired, "fill" or "auto"
        'samplepoints', # default: 8; number of points used to represent ellipses
                        # and circles on output
        'searchsize', # default: 30; maximum edges with negative cut values to check
                      # when looking for a minimum one during network simplex
        'size', # maximum drawing size, in inches
        'style', # graphics options, e.g. "filled" for clusters
        'URL', # URL associated with graph (format-dependent)

        # maintained for backward compatibility or rdot internal
        'layerseq'
    ]

    # Ancestor of Edge, Node, and Graph.
    #
    class Element

      attr_accessor :name, :options

      def initialize(params = {}, option_list = [])
        @name = params['name'] ? params['name'] : nil
        @options = {}

        option_list.each do |i|
          @options[i] = params[i] if params[i]
        end
      end

      private

      # Returns the string given in _id_ within quotes if necessary. Special
      # characters are escaped as necessary.
      #
      def quote_ID(id)
        # Ensure that the ID is a string.
        id = id.to_s

        # Return the ID verbatim if it looks like a name, a number, or HTML.
        return id if id =~ /\A([[:alpha:]_][[:alnum:]_]*|-?(\.[[:digit:]]+|[[:digit:]]+(\.[[:digit:]]*)?)|<.*>)\Z/m && id[-1] != ?\n

        # Return a quoted version of the ID otherwise.
        '"' + id.gsub('\\', '\\\\\\\\').gsub('"', '\\\\"') + '"'
      end

      # Returns the string given in _label_ within quotes if necessary. Special
      # characters are escaped as necessary. Labels get special treatment in
      # order to handle embedded *\n*, *\r*, and *\l* sequences which are copied
      # into the new string verbatim.
      #
      def quote_label(label)
        # Ensure that the label is a string.
        label = label.to_s

        # Return the label verbatim if it looks like a name, a number, or HTML.
        return label if label =~ /\A([[:alpha:]_][[:alnum:]_]*|-?(\.[[:digit:]]+|[[:digit:]]+(\.[[:digit:]]*)?)|<.*>)\Z/m && label[-1] != ?\n

        # Return a quoted version of the label otherwise.
        '"' + label.split(/(\\n|\\r|\\l)/).collect do |part|
          case part
            when "\\n", "\\r", "\\l"
              part
            else
              part.gsub('\\', '\\\\\\\\').gsub('"', '\\\\"').gsub("\n", '\\n')
          end
        end.join + '"'
      end
    end


    # Ports are used when a Node instance has its `shape' option set to
    # _record_ or _Mrecord_. Ports can be nested.
    #
    class Port
      attr_accessor :name, :label, :ports

      # Create a new port with either an optional name and label or a set of
      # nested ports.
      #
      # :call-seq:
      #   new(name = nil, label = nil)
      #   new(ports)
      #
      # A +nil+ value for +name+ is valid; otherwise, it must be a String or it
      # will be interpreted as +ports+.
      #
      def initialize(name_or_ports = nil, label = nil)
        if name_or_ports.nil? || name_or_ports.kind_of?(String)
          @name  = name_or_ports
          @label = label
          @ports = nil
        else
          @ports = name_or_ports
          @name  = nil
          @label = nil
        end
      end

      # Returns a string representation of this port. If ports is a non-empty
      # Enumerable, a nested ports representation is returned; otherwise, a
      # name-label representation is returned.
      #
      def to_s
        if @ports.nil? || @ports.empty?
          n = (name.nil? || name.empty?) ? '' : "<#{name}>"
          n + ((n.empty? || label.nil? || label.empty?) ? '' : ' ') + label.to_s
        else
          '{' + @ports.collect { |p| p.to_s }.join(' | ') + '}'
        end
      end
    end

    # A node representation. Edges are drawn between nodes. The rendering of a
    # node depends upon the options set for it.
    #
    class Node < Element

      attr_accessor :ports

      # Creates a new Node with the _params_ Hash providing settings for all
      # node options. The _option_list_ parameter restricts those options to the
      # list of valid names it contains. The exception to this is the _ports_
      # option which, if specified, must be an Enumerable containing a list of
      # ports.
      #
      def initialize(params = {}, option_list = NODE_OPTS)
        super(params, option_list)
        @ports = params['ports'] ? params['ports'] : []
      end

      # Returns a string representation of this node which is consumable by the
      # graphviz tools +dot+ and +neato+. The _leader_ parameter is used to indent
      # every line of the returned string, and the _indent_ parameter is used to
      # additionally indent nested items.
      #
      def to_s(leader = '', indent = '    ')
        label_option = nil

        if @options['shape'] =~ /^M?record$/ && !@ports.empty?
          # Ignore the given label option in this case since the ports should each
          # provide their own name/label.
          label_option = leader + indent + "#{quote_ID('label')} = #{quote_ID(@ports.collect { |port| port.to_s }.join(" | "))}"
        elsif @options['label']
          # Otherwise, use the label when given one.
          label_option = leader + indent + "#{quote_ID('label')} = #{quote_label(@options['label'])}"
        end

        # Convert all the options except `label' and options with nil values
        # straight into name = value pairs. Then toss out any resulting nil
        # entries in the final array.
        stringified_options = @options.collect do |name, val|
          unless name == 'label' || val.nil?
            leader + indent + "#{quote_ID(name)} = #{quote_ID(val)}"
          end
        end.compact

        # Append the specially computed label option.
        stringified_options.push(label_option) unless label_option.nil?

        # Join them all together.
        stringified_options = stringified_options.join(",\n")

        # Put it all together into a single string with indentation and return the
        # result.
        if stringified_options.empty?
          leader + quote_ID(@name) unless @name.nil?
        else
          leader + (@name.nil? ? '' : quote_ID(@name) + " ") + "[\n" +
              stringified_options + "\n" +
              leader + "]"
        end
      end

    end # class Node

    # A graph representation. Whether or not it is rendered as directed or
    # undirected depends on which of the programs *dot* or *neato* is used to
    # process and render the graph.
    #
    class Graph < Element

      # Creates a new Graph with the _params_ Hash providing settings for all
      # graph options. The _option_list_ parameter restricts those options to the
      # list of valid names it contains. The exception to this is the _elements_
      # option which, if specified, must be an Enumerable containing a list of
      # nodes, edges, and/or subgraphs.
      #
      def initialize(params = {}, option_list = GRAPH_OPTS)
        super(params, option_list)
        @elements   = params['elements'] ? params['elements'] : []
        @dot_string = 'graph'
      end

      # Calls _block_ once for each node, edge, or subgraph contained by this
      # graph, passing the node, edge, or subgraph to the block.
      #
      # :call-seq:
      #   graph.each_element {|element| block} -> graph
      #
      def each_element(&block)
        @elements.each(&block)
        self
      end

      # Adds a new node, edge, or subgraph to this graph.
      #
      # :call-seq:
      #   graph << element -> graph
      #
      def <<(element)
        @elements << element
        self
      end

      alias push <<

      # Removes the most recently added node, edge, or subgraph from this graph
      # and returns it.
      #
      # :call-seq:
      #   graph.pop -> element
      #
      def pop
        @elements.pop
      end

      # Returns a string representation of this graph which is consumable by the
      # graphviz tools +dot+ and +neato+. The _leader_ parameter is used to indent
      # every line of the returned string, and the _indent_ parameter is used to
      # additionally indent nested items.
      #
      def to_s(leader = '', indent = '    ')
        hdr = leader + @dot_string + (@name.nil? ? '' : ' ' + quote_ID(@name)) + " {\n"

        options = @options.to_a.collect do |name, val|
          unless val.nil?
            if name == 'label'
              leader + indent + "#{quote_ID(name)} = #{quote_label(val)}"
            else
              leader + indent + "#{quote_ID(name)} = #{quote_ID(val)}"
            end
          end
        end.compact.join("\n")

        elements = @elements.collect do |element|
          element.to_s(leader + indent, indent)
        end.join("\n\n")

        hdr + (options.empty? ? '' : options + "\n\n") +
            (elements.empty? ? '' : elements + "\n") + leader + "}"
      end

    end # class Graph

    # A digraph is a directed graph representation which is the same as a Graph
    # except that its header in dot notation has an identifier of _digraph_
    # instead of _graph_.
    #
    class Digraph < Graph

      # Creates a new Digraph with the _params_ Hash providing settings for all
      # graph options. The _option_list_ parameter restricts those options to the
      # list of valid names it contains. The exception to this is the _elements_
      # option which, if specified, must be an Enumerable containing a list of
      # nodes, edges, and/or subgraphs.
      #
      def initialize(params = {}, option_list = GRAPH_OPTS)
        super(params, option_list)
        @dot_string = 'digraph'
      end

    end # class Digraph

    # A subgraph is a nested graph element and is the same as a Graph except
    # that its header in dot notation has an identifier of _subgraph_ instead of
    # _graph_.
    #
    class Subgraph < Graph

      # Creates a new Subgraph with the _params_ Hash providing settings for
      # all graph options. The _option_list_ parameter restricts those options to
      # list of valid names it contains. The exception to this is the _elements_
      # option which, if specified, must be an Enumerable containing a list of
      # nodes, edges, and/or subgraphs.
      #
      def initialize(params = {}, option_list = GRAPH_OPTS)
        super(params, option_list)
        @dot_string = 'subgraph'
      end

    end # class Subgraph

    # This is an undirected edge representation.
    #
    class Edge < Element

      # A node or subgraph reference or instance to be used as the starting point
      # for an edge.
      attr_accessor :from

      # A node or subgraph reference or instance to be used as the ending point
      # for an edge.
      attr_accessor :to

      # Creates a new Edge with the _params_ Hash providing settings for all
      # edge options. The _option_list_ parameter restricts those options to the
      # list of valid names it contains.
      #
      def initialize(params = {}, option_list = EDGE_OPTS)
        super(params, option_list)
        @from = params['from'] ? params['from'] : nil
        @to   = params['to'] ? params['to'] : nil
      end

      # Returns a string representation of this edge which is consumable by the
      # graphviz tools +dot+ and +neato+. The _leader_ parameter is used to indent
      # every line of the returned string, and the _indent_ parameter is used to
      # additionally indent nested items.
      #
      def to_s(leader = '', indent = '    ')
        stringified_options = @options.collect do |name, val|
          unless val.nil?
            leader + indent + "#{quote_ID(name)} = #{quote_ID(val)}"
          end
        end.compact.join(",\n")

        f_s = @from || ''
        t_s = @to || ''

        if stringified_options.empty?
          leader + quote_ID(f_s) + ' ' + edge_link + ' ' + quote_ID(t_s)
        else
          leader + quote_ID(f_s) + ' ' + edge_link + ' ' + quote_ID(t_s) + " [\n" +
              stringified_options + "\n" +
              leader + "]"
        end
      end

      private

      def edge_link
        '--'
      end

    end # class Edge

    # A directed edge representation otherwise identical to Edge.
    #
    class DirectedEdge < Edge

      private

      def edge_link
        '->'
      end

    end # class DirectedEdge

  end # module DOT

end # module RGL
